### ISSUE TITLE
Datatable: filter facet does not work with p:columns

### ISSUE THREAD

**Issue by tuerker** [primefaces/primefaces#6591]
@Rapster After the recent changes on the datatable component (< 10 days) something is broken w.r.t. filtering.

Since several years, I have a lazy datatable where is based on p:columns together with a filter facet. This filter is not working anymore. When entering the load method of the LazyDataModel

load(int first, int pageSize, Map<String, SortMeta> sortBy, Map<String, FilterMeta> filterBy) 

the filterBy argument is not null but the hashmap is empty.

When i use same inner code fragment of the column separated over several p:column elements everything is ok. 

Please forgive that I am not able to provide here a reproducer for this non-trivial case. I hope you got my point and have an idea where to search for the bug.

**Comment by Rapster** [comment]
Fixed

**Comment by tuerker** [comment]
Dear @Rapster unfortunately the bug is still there. At least it not solved in the latest snapshot. Or is this fix not merged into the master?

**Comment by Rapster** [comment]
Please provide at least a XHTML sample

**Comment by tuerker** [comment]
I stripped a lot of the code. Hopefully you get an idea.

The filter of the following column with a filter facet is working:

```
  <p:column
    sortBy="#{entity.status}"
    filterBy="#{entity.status}"
    filterMatchMode="exact">
    <f:facet name="header">
      <h:outputText value="#{messages.statusLabel} x" />
    </f:facet>
    <f:facet name="filter">
      <h:selectOneMenu
        id="xstatus"
        onchange="PF('#{tableContext}#{tableType}').filter()">
        <f:selectItem itemLabel="#{messages.noFilterLabel}" />
        <f:selectItems
          value="#{SampleStatusEnum.values()}"
          var="item"
          itemLabel="#{item.label}" />
      </h:selectOneMenu>
    </f:facet>
    <h:outputText value="#{entity.status}" />
  </p:column>
```

while the "same" column within a p:columns widget is not working:

 ```
 <p:columns
    value="..."
    var="sampleAttributeEnum"
    filterable="#{not sampleAttributeEnum.annotationTypeMultiValued}"
    filterBy="#{entity[sampleAttributeEnum.name]}"
    sortable="#{not sampleAttributeEnum.annotationTypeMultiValued}"
    sortBy="#{entity[sampleAttributeEnum.name]}">
    <f:facet name="header">
      <h:outputText value="#{sampleAttributeEnum.labelWithUnit}" />
    </f:facet>
    <f:facet name="filter">
      <h:selectOneMenu
        id="enumFilter"
        rendered="#{sampleAttributeEnum.enumType}"
        onchange="PF('#{tableContext}#{tableType}').filter()"
        converter="omnifaces.GenericEnumConverter">
        <f:selectItem itemLabel="#{messages.noFilterLabel}" />
        <f:selectItems
          value="#{sampleAttributeEnum.getEnumValues(false)}"
          var="item"
          itemLabel="#{item.label}" />
      </h:selectOneMenu>
    </f:facet>
    ...
  </p:columns>
```

Using the filter of the column "status x" created via p:column works as you see in the next screenshot:

![Screenshot from 2020-12-03 15-00-53](https://user-images.githubusercontent.com/10051373/101027562-70380800-3578-11eb-9897-d8f33d3019c0.png)

Using the filter of the "same" column "status" created via p:columns does not work. see in the next screenshot:

![Screenshot from 2020-12-03 15-01-26](https://user-images.githubusercontent.com/10051373/101027581-76c67f80-3578-11eb-82d8-b48f95e8cf29.png)

After your fix, the load method of the LazyDataModel is even not invoked anymore. Formerly, it was invoked with an empty filterBy argument.

**Comment by MatthiasScherrerFGCZ** [comment]
@Rapster Additional information (before the latest fix):

- The filterBy was empty, i.e., 0 instead of the expected size 1.
- To be sure that the filterBy is correctly initialised, we simply did filter for another column without a face filter that is generated in the same p:columns. So after applying the first filter, the size of filterBy was 1 as expected. Then we filtered the column with the facet filter and the filterBy size was still 1 instead of the expected 2. We did this forth and back with various columns and the generated columns with facet filter always showed this behaviour.
- So somewhere along the processing of the filter and updating the filterBy, the facet filter is not processed.
- We use p:columns in our project a few times but only at one place with a facet filter.

**Comment by Rapster** [comment]
I'll have a look

**Comment by Rapster** [comment]
At this point, there is too much factors to take account for me to try. Please provide a reproducer. I tried with this sample, and not facing any issue:

```xml
<p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10" id="carTable" widgetVar="carsTable">

    <p:columns value="#{dtLazyView.columns}" var="column" headerText="#{column.header}" field="#{column.property}">
        <f:facet name="filter">
            <p:inputText onchange="PF('carsTable').filter()" />
        </f:facet>
    </p:columns>

</p:dataTable>
```

**Comment by MatthiasScherrerFGCZ** [comment]
@Rapster 
Yes, I see. We definitely have to provide a reproducer. Thanks.

**Comment by tuerker** [comment]
dear @Rapster could you please attach your reproducer such that we can adapt it. i could imagine that the problem is caused by using a selectonemenu in the facet filter. many thanks in advance.

**Comment by Rapster** [comment]
My sample is based on showcase (i just assembled dynamic and lazy showcase together)

**Comment by Rapster** [comment]
Also make sure you're using the last snapshot...

**Comment by tuerker** [comment]
@Rapster finally i managed to create a reproducer. 

[p-dynamiclazy.zip](https://github.com/primefaces/primefaces/files/5639170/p-dynamiclazy.zip)

when the columns are created via p:columns while the facet filter is conditionally (i.e., bound to a rendered condition), then this problem occurs. In the reproducer, i am just rendering the facet filter (selectonemenu) for the "brand" column only.

If you drop the rendered attribute, the filter is working well.

so a bug must have been introduced with the recent changes because our code was running for years. it is currently running with the snapshot of 2020 Nov 19.

**Comment by Rapster** [comment]
Thank you for providing a reproducer. Hope this time is the right one 😉

**Comment by tuerker** [comment]
Great. Problem is fixed now. Many thanks again. I really appreciate that you all are so responsive.

**Comment by MatthiasScherrerFGCZ** [comment]
@Rapster Thanks :)

**Comment by Rapster** [comment]
Forgot to mention, since sortBy & filterBy are the same, you can simplify this by using `UIColumns#field`:

```xml
<p:columns
   value="..."
   var="sampleAttributeEnum"
   filterable="#{not sampleAttributeEnum.annotationTypeMultiValued}"
   sortable="#{not sampleAttributeEnum.annotationTypeMultiValued}"
   field="#{entity[sampleAttributeEnum.name]}">

   // some stuff

 </p:columns>
```

**Comment by tuerker** [comment]
Thanks for the hint. Actually it works as you stated for p:columns. 

However, you cannot do the same for p:column. There you will get the following error:

`Column 'field' attributes must be non null.`

**Comment by tuerker** [comment]
just replace in the reproducer above the html file to verify this:

```
  <h:form id="form">
    <p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10">
      <p:column
        field="#{car.brand}"
        filterMatchMode="contains">
        <f:facet name="header">
          <h:outputText value="brand" />
        </f:facet>
        <h:outputText value="#{car.brand}" />
      </p:column>
    </p:dataTable>
  </h:form>

```

while this works:

```
  <h:form id="form">
    <p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10">
      <p:column
        sortBy="#{car.brand}"
        filterBy="#{car.brand}"
        filterMatchMode="contains">
        <f:facet name="header">
          <h:outputText value="brand" />
        </f:facet>
        <h:outputText value="#{car.brand}" />
      </p:column>
    </p:dataTable>
  </h:form>

```

**Comment by Rapster** [comment]
`UIColumn#field` expects a constant

**Comment by tuerker** [comment]
i see.

**Comment by tuerker** [comment]
i can confirm that the following works:

  ```
<h:form id="form">
    <p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10">
      <p:column
        field="brand"
        filterMatchMode="contains">
        <f:facet name="header">
          <h:outputText value="brand" />
        </f:facet>
        <h:outputText value="#{car.brand}" />
      </p:column>
    </p:dataTable>
  </h:form>

```

**Comment by tuerker** [comment]
thanks again.

**Comment by tuerker** [comment]
i have detected that it this approach gets trickier with at least lazy model when the column is based on embedded values.

This is the currently running column:

```
  <p:column
    sortBy="#{entity.address.zip}"
    filterBy="#{entity.address.zip}"
    filterMatchMode="contains">
    <f:facet name="header">
      <h:outputText value="#{messages.zipLabel}" />
    </f:facet>
    <h:outputText value="#{entity.address.zip}" />
  </p:column>

```

When changed to the field approach it will not run anymore:
```
  <p:column
     field="zip"
    filterMatchMode="contains">
    <f:facet name="header">
      <h:outputText value="#{messages.zipLabel}" />
    </f:facet>
    <h:outputText value="#{entity.address.zip}" />
  </p:column>

```

The following instead will run since the query then will mapped again correctly:
```
  <p:column
     field="address.zip"
    filterMatchMode="contains">
    <f:facet name="header">
      <h:outputText value="#{messages.zipLabel}" />
    </f:facet>
    <h:outputText value="#{entity.address.zip}" />
  </p:column>

```

**Comment by Rapster** [comment]
Assuming var="entity", try field="address.zip"?

**Comment by tuerker** [comment]
exactly this runs. see third code fragment above. thanks.

**Comment by Rapster** [comment]
My bad I didn't see, then it works as expected

**Comment by tuerker** [comment]
would be great to generalize the column concept one step ahead to avoid redundancy problems by just writing

```
  <p:column
    field="address.country.name">
    <f:facet name="header">
      <h:outputText value="#{messages.countryLabel}" />
    </f:facet>
  </p:column>

```
instead of 
```

  <p:column
    field="address.country.name">
    <f:facet name="header">
      <h:outputText value="#{messages.countryLabel}" />
    </f:facet>
    <h:outputText value="#{entity.address.country.name}" />
  </p:column>

```

thereby, the field defines the output the table field.

**Comment by Rapster** [comment]
Weird, first sample should work

**Comment by tuerker** [comment]
great. i had not tested this. it was an immediate wish after seeing that field is available.

**Comment by tuerker** [comment]
it is actually working. many thanks again.

**Comment by tuerker** [comment]
i have detected a case where it is not working:

```
  <p:column
    field="title"
    filterMatchMode="contains">
    <f:facet name="header">
      <h:outputText value="#{messages.titleLabel}" />
    </f:facet>
  </p:column>
```

while the following is working.

```
  <p:column
    field="title"
    filterMatchMode="contains">
    <f:facet name="header">
      <h:outputText value="#{messages.titleLabel}" />
    </f:facet>
    <h:outputText value="#{entity.title}" />
  </p:column>
```

could it be that "title" is somehow reserved?

**Comment by tuerker** [comment]
the following error is thrown:

```
java.lang.NullPointerException: Argument Error: Parameter text is null
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeText(HtmlResponseWriter.java:955)
	at org.primefaces.component.datatable.DataTableRenderer.encodeCell(DataTableRenderer.java:1350)
	at org.primefaces.component.datatable.DataTableRenderer.encodeRow(DataTableRenderer.java:1265)
	at org.primefaces.component.datatable.DataTableRenderer.encodeRows(DataTableRenderer.java:1157)
	at org.primefaces.component.datatable.DataTableRenderer.encodeTbody(DataTableRenderer.java:1094)
	at org.primefaces.component.datatable.DataTableRenderer.encodeTbody(DataTableRenderer.java:1043)
	at org.primefaces.component.datatable.DataTableRenderer.encodeRegularTable(DataTableRenderer.java:386)
	at org.primefaces.component.datatable.DataTableRenderer.encodeMarkup(DataTableRenderer.java:314)
	at org.primefaces.component.datatable.DataTableRenderer.encodeEnd(DataTableRenderer.java:94)
```

**Comment by tuerker** [comment]
This might be related with the other issue which you have fixed but i do not have at the moment since it is not in the master.

**Comment by Rapster** [comment]
I fixed that issue same time as #6609
